### ISSUE TITLE
React.Lazy component with other exports do not get refreshed

### ISSUE THREAD

**Issue by MattyBalaam** [pmmmwh/react-refresh-webpack-plugin#277]
If I have a component on the default export loaded lazily I am unable to see any refreshes if I also have any named exports.

`const Component = lazy(() => import('./AComponent'))`

The same component import normally will refresh as expected.

`import Component from ./AComponent'`

```
export const myNamedExport = {
  one: 'one'
}

export default function AComponent() {
  return <p>hello</p>;
}
```

**Comment by MattyBalaam** [comment]
Here is an example

[Archive.zip](https://github.com/pmmmwh/react-refresh-webpack-plugin/files/5690610/Archive.zip)

**Comment by MattyBalaam** [comment]
I’m digging around a little (very crudely). I can see that without the named export `getModuleExports` in `RefreshUtils` returns a object, but after adding one named export it returns an array of 6 entries.

![Uploading Screenshot 2020-12-15 at 07.54.42.png…]()

**Comment by ShravanSunder** [comment]
@MattyBalaam  did you find out what was happening?  I have a situation where fast refresh breaks when i use Lazy

**Comment by ShravanSunder** [comment]
Conditions:
1. componentA has any named exports and defaults
2. componentA is imported via Lazy 

In this situation seems like fast refresh stops working

**Comment by MattyBalaam** [comment]
I didn’t feel I was really understanding what was going on, so I bailed out of my exploration, but hopefully the test case and info I have left will help someone else work out a solution.

**Comment by pmmmwh** [comment]
Sorry for chiming in late.

I suspect this actually have something to do with the actual return value of a dynamically imported `lazy` component, will do a bit of digging this weekend and see if I can find out whats wrong.

**Comment by frehner** [comment]
Ah I just ran into this as well - when using a microfrontend setup, you almost never have a default export from another MFE:

```js
import {Button} from 'my-styleguide'
```

(it wouldn't make sense for the `my-styleguide` MFE to have a default export)

I'll dig a bit and see if I can figure something out or not

**Comment by frehner** [comment]
Digging in, and I may have an idea of what's going on. I could be wrong though 🙂 

The runtime [checks to see if the currentExports are a refresh boundary](https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/main/loader/RefreshModule.runtime.js#L21) before it applies any updates. 

When there are multiple exports, the function that checks that only returns true [if each and every export is a React component](https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/3bcc72b63e9f0b3bbfb26339c7974295cd84e7af/lib/runtime/RefreshUtils.js#L107)

(For reference, [here's the `react-refresh/runtime` code that determines if an export is a React component](https://github.com/facebook/react/blob/6d94017c4287195d8d6c5173ea663285820e9f61/packages/react-refresh/src/ReactFreshRuntime.js#L671-L720))  

As an additional note, the reason there is never an error message that HMR didn't actually work is because [this line comes out to false](https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/3bcc72b63e9f0b3bbfb26339c7974295cd84e7af/loader/RefreshModule.runtime.js#L82) and thus `module.hot.invalidate()` is never called.

So, I guess some questions:

1. Is there a technical reason that every export has to be a React component for refresh to work?
2. Is there a way that we could either improve the way that react-refresh determines if an export is a React component, or optionally... trick it into trusting us that it's a React component? (e.g. it's a promise that resolves to a component, which is used in `React.lazy`, so trust me React! 😄 )
3. Is there a better way to handle the `else` statement there, so that `module.hot.invalidate()` is called and lets us know that HMR didn't actually work?

**Comment by pmmmwh** [comment]
> Is there a technical reason that every export has to be a React component for refresh to work?

React components are known to be side-effects free. That assumption doesn't hold outside.

> Is there a way that we could either improve the way that react-refresh determines if an export is a React component, or optionally... trick it into trusting us that it's a React component? (e.g. it's a promise that resolves to a component, which is used in React.lazy, so trust me React! 😄 )

Can't really comment on the first part of the question, but if you name your exports accordingly (Note: **proceed with caution, not recommended**).

> Is there a better way to handle the else statement there, so that module.hot.invalidate() is called and lets us know that HMR didn't actually work?

If you remove the check on `prevExports`, does it bail out correctly? I think we could remove that part without compromising on the soundness of the runtime.

**Comment by frehner** [comment]
>React components are known to be side-effects free. That assumption doesn't hold outside.

That makes sense. Because I'm not very familiar with HMR, how granular could you potentially get with it? For example, is it possible to check each export, see if it's a React component or not, and do fast refresh if _only React components_ were updated? Or is that not possible?

>If you remove the check on prevExports, does it bail out correctly? I think we could remove that part without compromising on the soundness of the runtime.

hm, it does appear to call invalidate if I remove that line, but there doesn't appear to be any error messages in the console and it appears as if HMR still worked correctly.

**Comment by frehner** [comment]
Hm, I discovered that my situation is actually different than this one and am trying to narrow it down again. 

(A minimal repo I was using helped me discover the causes above, but it turns out my app's problems were actually different than the minimal repo's problems. 🙂 )

**Comment by pmmmwh** [comment]
Should be fixed in `0.5.0-beta.1`.
